src/test/java/software/amazon/awssdk/crt/test/EventStreamClientConnectionTest.java

```diff
@@ -0,0 +1,674 @@
+        SocketOptions socketOptions = new SocketOptions();
+        socketOptions.connectTimeoutMs = 3000;
+        socketOptions.domain = SocketOptions.SocketDomain.IPv4;
+        socketOptions.type = SocketOptions.SocketType.STREAM;
+
+        EventLoopGroup elGroup = new EventLoopGroup(1);
+        ServerBootstrap bootstrap = new ServerBootstrap(elGroup);
+        ClientBootstrap clientBootstrap = new ClientBootstrap(elGroup, null);
+        final boolean[] connectionReceived = {false};
+        final boolean[] connectionShutdown = {false};
+
+        ServerListener listener = new ServerListener("127.0.0.1", (short)8033, socketOptions, null, bootstrap, new ServerListenerHandler() {
+            private ServerConnectionHandler connectionHandler = null;
+
+            public ServerConnectionHandler onNewConnection(ServerConnection serverConnection, int errorCode) {
+                connectionReceived[0] = true;
+
+                connectionHandler = new ServerConnectionHandler(serverConnection) {
+
+                    @Override
+                    protected void onProtocolMessage(List<Header> headers, byte[] payload, MessageType messageType, int messageFlags) {
+                    }
+
+                    @Override
+                    protected ServerConnectionContinuationHandler onIncomingStream(ServerConnectionContinuation continuation, String operationName) {
+                        return null;
+                    }
+                };
+
+                return connectionHandler;
+            }
+
+            public void onConnectionShutdown(ServerConnection serverConnection, int errorCode) {
+                connectionShutdown[0] = true;
+                connectionHandler.close();
+            }
+        });
+
+        final boolean[] clientConnected = {false};
+        final boolean[] clientConnectionClosed = {false};
+        final ClientConnection[] clientConnectionArray = {null};
+
+        CompletableFuture<Void> connectFuture = ClientConnection.connect("127.0.0.1", (short)8033, socketOptions, null, clientBootstrap, new ClientConnectionHandler() {
+            @Override
+            protected void onConnectionSetup(ClientConnection connection, int errorCode) {
+                clientConnected[0] = true;
+                clientConnectionArray[0] = connection;
+            }
+
+            @Override
+            protected void onProtocolMessage(List<Header> headers, byte[] payload, MessageType messageType, int messageFlags) {
+
+            }
+
+            @Override
+            protected void onConnectionClosed(int closeReason) {
+                clientConnectionClosed[0] = true;
+                this.close();
+            }
+        });
+
+        connectFuture.get();
+        assertNotNull(clientConnectionArray[0]);
+        clientConnectionArray[0].closeConnection(0);
+        // this should be damn near instant, but give it a second just to be safe.
+        Thread.sleep(1000);
```

Comment:
Can we not drive this with a future that gets finished in `onConnectionClosed()`?

I'm a little worried about these sporadically failing in CI because the sleep wasn't long enough on multi-tenant boxes.

Reply:
This applies to all the sleeping tests.